app/src/main/java/com/arctouch/codechallenge/data/MoviesDataSource.java

```diff
@@ -0,0 +1,33 @@
+package com.arctouch.codechallenge.data;
+
+import androidx.paging.PageKeyedDataSource;
+
+import com.arctouch.codechallenge.data.model.Genre;
+import com.arctouch.codechallenge.data.model.GenreResponse;
+import com.arctouch.codechallenge.data.model.Movie;
+import com.arctouch.codechallenge.data.model.UpcomingMoviesResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.reactivex.Observable;
+
+public abstract class MoviesDataSource /*extends PageKeyedDataSource<Integer, Movie>*/ {
+    //public abstract Observable<UpcomingMoviesResponse> subscribeForAllmovies(String language, int page, String region);
+    public abstract Observable<List<Movie>> subscribeForAllmovies(String language, int page, String region);
+    public abstract Observable<List<Movie>> subscribeForMoviesByTitle(String name, int page);
+    public abstract Observable<GenreResponse> subscribeForGenres(String language);
+
+    public List<Movie> applyGenresToMovies(UpcomingMoviesResponse moviesResponse, GenreResponse genreResponse){
```

Comment:
AndroidStudio gives a warning because this method could be protected. Do you think that warning is valid? Would it improve your code if you made that change?

Reply:
In fact It is valid, And doing would improve the encapsulation of my code.